cmd/podman/networks/create.go

```diff
@@ -52,6 +54,8 @@ func networkCreateFlags(cmd *cobra.Command) {
 	flags.IPSliceVar(&networkCreateOptions.Gateways, gatewayFlagName, nil, "IPv4 or IPv6 gateway for the subnet")
 	_ = cmd.RegisterFlagCompletionFunc(gatewayFlagName, completion.AutocompleteNone)
 
+	flags.BoolVar(&networkCreateOptions.NoAutoGateway, "no-auto-gateway", false, "do not automatically add gateways to subnets")
+
```

Comment:
I am not a fan of this flag. Can you explain the use case more? Compared to e.g. --internal?
I really do not want to add more top level options to the Network type, if this is truly needed I think it is a much better fit via the generic options map.

Reply:
On macvlan it does the exact same thing an `--internal`. I confirmed that all internal does on macvlan is prevent a default gateway. from being added automatically.

On bridge is where the difference comes in. Here, `--internal` also applies firewall rules that prevent external communication. So let's say you don't want a default route but add static routes to certain destinations. You can't achieve that with `--internal` as the firewall rules will block communication to those destinations.

Now that I write this, I think it might be possible to simply have the firewall rules allow communication to those static route destinations. Although that makes the meaning of `--internal` a little murky. In that case `--internal` would be internal except for those allowed packets to static route destinations.

Reply:
--internal does not add firewall rules to prevent external communication with bridge, we just set a sysctl to block routing on the bridge interface so it is impossible to reach ips outside the subnet.

I think having this as a use case is fine, but it should not be top level. Just put it in the existing `--opt` which just maps to a `[string]string` map, so something like `--opt no_auto_gateway=1` could be used.

We shouldn't change the meaning of internal at this point.

Reply:
Sounds good

Reply:
A bridge network is a NAT router between the host and the pod/container. This NAT router uses the host route table. To make this work a default gateway for the bridge network is necessary.

On a macvlan network the gateway is explicitly defined with the --gateway option. We don't need no_auto_gateway for macvlan.

Combining bridge and macvlan in 1 pod/container is complex and requires rule-based routing. But this is complex and out of scope.

I only see multiple networks per pod/container with macvlan networks. I wouldn't allow a combination of macvlan and bridge networks. Therefore the no_auto_gateway flag is not needed.

Reply:
The no_auto_gateway flag in this PR will be replaced by no_default_gateway. The relevant PR is containers/netavark#697





> A bridge network is a NAT router between the host and the pod/container. This NAT router uses the host route table. To make this work a default gateway for the bridge network is necessary.

For a bridge network a default route is not necessary if you have a relevant static route set up via the gateway.


> On a macvlan network the gateway is explicitly defined with the --gateway option. We don't need no_auto_gateway for macvlan.

On macvlan even without the --gateway option a gateway is added to the network config.

Reply:
Edited the above comment, because I wrote no_default_gateway instead of no_default_route.

Reply:
> The no_auto_gateway flag in this PR will be replaced by no_default_route. The relevant PR is [containers/netavark#697](https://github.com/containers/netavark/pull/697)
> 
> > A bridge network is a NAT router between the host and the pod/container. This NAT router uses the host route table. To make this work a default gateway for the bridge network is necessary.
> 
> For a bridge network a default route is not necessary if you have a relevant static route set up via the gateway.

You can create static routes if you know the gateway address automatically defined by the bridge driver.
In a normal IP network, packets are allowed to follow different routes. Only the destination determines the route.
This does not apply to NAT routing. For connections that go over NAT, all packets must go through the same NAT router (because of connection tracking). This can be achieved with rule-based routing. However, this is out of scope I think.
Therefore I wouldn't recommend using bridge and macvlan in the pod/container.

> 
> > On a macvlan network the gateway is explicitly defined with the --gateway option. We don't need no_auto_gateway for macvlan.
> 
> On macvlan even without the --gateway option a gateway is added to the network config.

If the gateway address is not defined what address is then used? It is not wise not to make assumptions about what the gateway address might be (there is no standard). If this is already done by the macvlan driver, than this is unexpected behavior in my opinion.

Reply:
To justify the feature in one (or two) sentence: If we give the user the ability to add routes to the container network namespace, then we should give him the ability to disable the otherwise automatically generated default route. That way he has full control over the routes in the network namespace if he desires.

Greetings from Delft, btw.

Reply:
I totally agree. multiple networks without static routes are almost useless. I don't like the no_default_path flag. But since podman assumes a gateway address without specifying one, I don't see any other solution.

We are researching podman for our internet gateways. A pod with 2 networks. 1 with a real internet ip number and 1 with an internal private ip address and containers with bind, nginx, postfix, asterisk, etc. So i hope this PR is in the next release :)

Podman also assumes a ip-range without specifying one (based on the subnet). This can result in duplicate IP addresses if you forget to specify an ip_address when creating the pod :-(.
It would be nice to have 'none' for --ip-range and --gateway.

I also miss a standard way to create firewall rules (nftable/iptables) for a pod.